common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java

```diff
@@ -209,4 +190,51 @@ private ShuffleMetrics() {
     }
   }
 
+  private class ManagedBufferIterator implements Iterator<ManagedBuffer> {
+
+    private int index = 0;
+    private String appId;
+    private String execId;
+    private String shuffleId;
+    // An array containing mapId and reduceId pairs.
+    private int[][] mapIdAndReduceIds;
```

Comment:
Actually, I mean a single array. e.g.

```
int[] mapIdAndReduceIds;

mapIdAndReduceIds = new int[blockIds.length * 2];
mapIdAndReduceIds[0] = mapId1;
mapIdAndReduceIds[1] = reduceId1;
mapIdAndReduceIds[2] = mapId2;
mapIdAndReduceIds[3] = reduceId2;
etc etc etc
```

Reason being that if you really have millions of these, each "child" array in your two-dimensional array wastes 16 (or 20?) bytes (16 bytes of object overhead + 4 bytes for the array length). Looking in jvisual vm, an empty array actually consumes 24 bytes, so it seems the JVM is aligning things and wasting an extra 4 bytes per array...